src/katsdpmodels/primary_beam.py

```diff
@@ -0,0 +1,256 @@
+        """Maximum distance from the pointing centre at which model has full coverage."""
+        raise NotImplementedError()
+
+    def max_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has any coverage."""
+        raise NotImplementedError()
+
+    @property
+    def is_circular(self) -> bool:
+        """Whether this model is circularly-symmetric about the pointing centre."""
+        raise NotImplementedError()
+
+    @property
+    def is_unpolarized(self) -> bool:
+        """Whether this model ignores polarization.
+
+        If true, it is guaranteed that the Jones matrices describing the beam are
+        scaled identity matrices.
+        """
+        raise NotImplementedError()
+
+    @property
+    def parameters(self) -> Sequence[Parameter]:
+        """Parameters of the model."""
+        raise NotImplementedError()
+
+    def sample(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noqa: E741
+               frame: Union[AltAzFrame, RADecFrame],
+               output_type: OutputType,
+               parameters: Dict[str, u.Quantity] = {}) -> np.ndarray:
+        """Sample the primary.
+
+        A sample is returned for each combination of a position (given by `l`,
+        `m`) with a frequency. The dimensions of the output will be first those
+        of `frequency`, then those of `l` and `m` (which are broadcast with
+        each other), and finally the row and column for matrices if
+        `output_type` is one of the matrix types.
+
+        Parameters
+        ----------
+        l
+            Horizontal coordinates (interpreted according to `frame`).
+        m
+            Vertical coordinates (interpreted according to `frame`).
+        frequency
+            Frequencies to sample
+        frame
+            Specifies how to interpret the coordinates.
+        output_type
+            The value to compute. It must be one of
+
+            OutputType.JONES_HV
+                Jones matrix with linear basis corresponding to horizontal and
+                vertical directions (see the class documentation for sign
+                conventions).
+            OutputType.JONES_XY
+                Jones matrix with linear basis corresponding to the IAU X
+                (north) and Y (east) directions on the celestial sphere.
+            OutputType.MUELLER
+                A 4x4 Mueller matrix describing the effect on each Stokes
+                parameter (IQUV), assuming that both antennas share the same
+                beam.
+            OutputType.UNPOLARISED_POWER
+                Power attenuation of unpolarized sources, assuming that both
+                antennas share the same beam. This is the same as the first
+                element of Mueller matrix above.
+        parameters
+            Additional parameters to the model. Parameters that are not used
+            are silently ignored.
+
```

Comment:
I'm going to add an optional `out` parameter that lets the user provide the memory. It looks like memory allocation/faulting is a significant part of the execution time so recycling memory should help.